packages/million/block.ts

```diff
@@ -0,0 +1,262 @@
+  // Inserts text nodes into the DOM at the correct position.
+  for (let i = 0, j = edits.length; i < j; ++i) {
+    const current = edits[i]!;
+    const initsLength = current.inits.length;
+    if (!initsLength) continue;
+    const el = getCurrentElement(current, root);
+    for (let k = 0; k < initsLength; ++k) {
+      const init = current.inits[k]!;
+      insertText(el, init.value, init.index);
+    }
+  }
+
+  return (
+    props?: Props | null,
+    key?: string,
+    shouldUpdate?: (oldProps: Props, newProps: Props) => boolean,
+  ) => {
+    return new Block(root, edits, props, key ?? props?.key, shouldUpdate);
+  };
+};
+
+export class Block extends AbstractBlock {
+  root: HTMLElement;
+  edits: Edit[];
+  // Cache for getCurrentElement()
+  cache = new Map<number, HTMLElement>();
+
+  constructor(
+    root: HTMLElement,
+    edits: Edit[],
+    props?: Props | null,
+    key?: string,
+    shouldUpdate?: (oldProps: Props, newProps: Props) => boolean,
+  ) {
+    super();
+    this.root = root;
+    this.props = props;
+    this.edits = edits;
+    this.key = key;
+    if (shouldUpdate) this.shouldUpdate = shouldUpdate;
+  }
+  mount(
+    parent?: HTMLElement,
+    refNode: Node | null = null,
+    inPlace = false,
+  ): HTMLElement {
+    if (this.el) return this.el;
+    // cloneNode(true) uses less memory than recursively creating new nodes
+    const root = cloneNode$.call(this.root, true) as HTMLElement;
+
+    for (let i = 0, j = this.edits.length; i < j; ++i) {
+      const current = this.edits[i]!;
+      const el = getCurrentElement(current, root, this.cache, i);
+      for (let k = 0, l = current.edits.length; k < l; ++k) {
+        const edit = current.edits[k]!;
+        const hasHole = 'hole' in edit && edit.hole;
+        const value = hasHole ? this.props![edit.hole!] : undefined;
+
+        if (edit.type === 'block') {
+          edit.block.mount(el, childNodes$.call(el)[edit.index]);
+        } else if (edit.type === 'child') {
+          if (value instanceof AbstractBlock) {
+            value.mount(el);
+            continue;
+          }
+          // insertText() on mount, setText() on patch
+          insertText(el, String(value), edit.index);
+        } else if (edit.type === 'event') {
+          const patch = createEventListener(
+            el,
+            edit.name,
+            // Events can be either a hole or a function
+            hasHole ? value : edit.listener,
+          );
+          if (hasHole) {
+            edit.patch = patch;
+          }
+        } else if (edit.type === 'attribute') {
+          setAttribute(el, edit.name, value);
+        } else if (edit.type === 'style') {
+          setStyleAttribute(el, edit.name, value);
+        } else {
+          setSvgAttribute(el, edit.name, value);
+        }
+      }
+    }
+
+    if (parent) {
+      if (inPlace) {
+        while (root.childNodes.length > 0) {
+          parent.appendChild(root.childNodes[0] as Node);
+        }
```

Comment:
Should we cache `childNodes` calls?

```ts
let rootChildNodes;
while ((rootChildNodes = root.childNodes).length > 0) {
  parent.appendChild(rootChildNodes[0] as Node);
}
```

And shouldn't we use cached prototype `childNodes$` here?

Also, I am wondering if `firstChild` is faster:

```ts
let rootFirstChild;
while ((rootFirstChild = root.firstChild)) {
  parent.appendChild(rootFirstChild);
}
```

Reply:
firstChild node access is def the faster way, thanks for the suggestion